File: solutions/erlang/secret-handshake/1/src/secret_handshake.erl
-module(secret_handshake).

-export([commands/1]).

%  1 - 00001 = wink
%  2 - 00010 = double blink
%  4 - 00100 = close your eyes
%  8 - 01000 = jump
% 16 - 10000 = Reverse the order of the operations in the secret handshake.

commands(Number) -> lists:reverse(command(Number)).

command(Number) when Number >= 16 -> lists:reverse(command(Number - 16));
command(Number) when Number >= 8 -> ["jump" | command(Number - 8)];
command(Number) when Number >= 4 -> ["close your eyes" | command(Number - 4)];
command(Number) when Number >= 2 -> ["double blink" | command(Number - 2)];
command(Number) when Number == 1 -> ["wink"];
command(_) -> [].

File: solutions/elixir/secret-handshake/1/lib/secret_handshake.ex
import Bitwise

defmodule SecretHandshake do
  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """


  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    actions = [
      {0b1, "wink"},
      {0b10, "double blink"},
      {0b100, "close your eyes"},
      {0b1000, "jump"}
    ]

    handshake =
      actions
      |> Enum.reduce([], fn {bit, action}, acc ->
        if (Bitwise.band(code, bit)) != 0, do: acc ++ [action], else: acc
      end)

    if (Bitwise.band(code, 0b10000)) != 0 do
      Enum.reverse(handshake)
    else
      handshake
    end
  end
end
